base/ltxref.dtx

```diff
@@ -258,10 +258,20 @@
 % As the commands are now protected we also need expandable versions for use
 % in \cs{ifthenelse}:
 %    \begin{macrocode}
-\def\@kernel@pageref@exp#1{\csname cs_if_exist:cTF\endcsname
-   {r@#1}{\csname tl_item:cn\endcsname{r@#1}{2}}{0}}
-\def\@kernel@ref@exp#1{\csname cs_if_exist:cTF\endcsname
-   {r@#1}{\csname tl_item:cn\endcsname{r@#1}{1}}{0}}
+\ExplSyntaxOn
+\cs_new:Npn \@kernel@pageref@exp #1
+  {
+    \cs_if_exist:cTF {r@#1}
+      { \exp_last_unbraced:Nv \use_ii:nn {r@#1} }
```

Comment:
This is easier than `\tl_item:cn`, but assumes that `\r@#1` will always contain two items. I believe that's also true when `hyperref` is loaded, right?

Reply:
No sorry this isn't correct. With hyperref there are five items and it will break there. This should be reverted. I don't mind the \ExplSyntaxOn/Off, but should one really switch that on and off only for two csnames?

Reply:
Ah, that's what I feared.  I used `expl3` syntax because it's less work and looks cleaner than building the csname (and if gradually we're moving to `expl3`, I don't see why not).

But mainly I think `\tl_item:nn` is (maybe unnecessarily) slow compared to `\use_<item>:nn` for a often-used command like `\ref`.  Maybe (regardless if `expl3` syntax or otherwise):
```latex
\cs_new:Npn \@kernel@ref@exp #1
  {
    \cs_if_exist:cTF {r@#1}
      { \exp_last_unbraced:Nv \@kernel@ref@get {r@#1} }
      { 0 }
  }
\cs_new:Npn \@kernel@ref@get #1 #2 {#1}
```
then `hyperref` can simply redefine `\cs_gset:Npn \@kernel@ref@get #1 #2 #3 #4 #5 {#1}`.

Reply:
> then `hyperref` can simply redefine `\cs_gset:Npn \@kernel@ref@get #1 #2 #3 #4 #5 {#1}`.

we should not encourage changes to code marked kernel by non-kernel packages (even if our own), it looks like as we would be promoting that.

Reply:
The name could be changed, of course, but yeah, there's that. `\tl_item:Nn` wouldn't have that problem, so I'm fine with that too.

Reply:
beside not encouraging redefinition of `@kernel` commands: I think we should also try to define things so that there is no need for external packages to redefine interna. The more redefinitions I can take out of hyperref the better. Also once we have decided about how to extend the reference system  the number/type of arguments should be unified too and then we  can exchange that with a faster version.

Reply:
Maybe use `\tl_head:e { \exp_args:Nv \tl_tail:n { r@#1 } }` instead?

Reply:
The `\tl_head:e/tl_tail:n` combo should be considerably faster than `\tl_item:n`, but if you're really after maximum performance here, the following should be even faster:

```
\cs_new:Npn \use_item_ii:n #1
  {
    \__kernel_exp_not:w
    \tex_expanded:D
      {
        \if_false: { \fi: \__use_item_ii_aux:w #1 }
      }
  }
\cs_new:Npn \__use_item_ii_aux:w #1#2
  {
    { \__kernel_exp_not:w {#2} }
    \exp_after:wN \use_none:n \exp_after:wN { \if_false: } \fi:
  }
```

Reply:
no I'm not after maximum performance, I like readable code more ;-). I'm okay with `\tl_head:v { r@#1 }` and `\tl_head:e { \tl_tail:v { r@#1 }}` for the first two (but imho it is already difficult see which item `\tl_head:e { \tl_tail:e { \tl_tail:e { \tl_tail:v { r@#1 } }}}` would extract.

Reply:
I'd not use multiple nested `\tl_tail:(n|e)` either :) And I understand the point about readable code (though I think at the top-level the `\use_item_ii:n` would be readable, and the implementation is just as cryptic as the implementation of `\tl_head:n` or `\tl_tail:n`; but this would be an otherwise useless addition to the kernel).